allennlp/modules/elmo.py

```diff
@@ -0,0 +1,422 @@
+import json
+
+from typing import Union, List, Dict, Any
+
+import torch
+from torch.autograd import Variable
+
+import numpy
+import h5py
+
+from overrides import overrides
+
+from allennlp.common.file_utils import cached_path
+from allennlp.common.checks import ConfigurationError
+from allennlp.common import Registrable, Params
+from allennlp.modules.elmo_lstm import ElmoLstm
+from allennlp.modules.highway import Highway
+from allennlp.modules.scalar_mix import ScalarMix
+from allennlp.nn.util import remove_sentence_boundaries, add_sentence_boundary_token_ids
+from allennlp.data import Vocabulary
+from allennlp.data.token_indexers.elmo_indexer import ELMoCharacterMapper
+
+# pylint: disable=attribute-defined-outside-init
+
+
+@Registrable.register('elmo')
+class Elmo(torch.nn.Module, Registrable):
+    """
+    Compute ELMo representations using a pre-trained bidirectional language model.
+
+    See "Deep contextualized word representations", Peters et al. for details.
+
+    This module takes character id input and computes ``num_elmo_layers`` different layers
+    of ELMo representations.  Typically ``num_elmo_layers`` is 1 or 2.  For example, in
+    the case of the SRL model in the above paper, ``num_elmo_layers=1`` where ELMo was included at
+    the input token representation layer.  In the case of the SQuAD model, ``num_elmo_layers=2``
+    as ELMo was also included at the GRU output layer.
+
+    In the implementation below, we learn separate scalar weights for each output layer,
+    but only run the biLM once on each input sequence for efficiency.
+
+    Parameters
+    ----------
+    options_file : str
+        ELMo JSON options file
+    weight_file : str
+        ELMo hdf5 weight file
+    num_elmo_layers: int
```

Comment:
Can we call this `num_scalar_mixture_outputs` or something? I would assume `num_elmo_layers` would mean the number of layers in the model, rather than  what it actually is.

Reply:
I'd like to leave lower level details about `ScalarMix` out of this, and it will make it easier to swap out the weighting if we discover a different a different weighting is better.  How about `num_output_layers`?

Reply:
I used `num_output_representations`